src/c++/perf_analyzer/test_dataloader.cc

```diff
@@ -673,4 +671,279 @@ TEST_CASE(
+        "If error reading output data file, an error is NOT raised, and "
+        "instead GetOutputData will return nullptr with a batch1_size of 0"))
+{
+  MockDataLoader dataloader;
+
+  std::shared_ptr<ModelTensorMap> inputs = std::make_shared<ModelTensorMap>();
+  std::shared_ptr<ModelTensorMap> outputs = std::make_shared<ModelTensorMap>();
+
+  ModelTensor output1 = TestDataLoader::CreateTensor("OUTPUT1");
+
+  std::string dir{"fake/path"};
+
+  SUBCASE("BYTES (string) data") { output1.datatype_ = "BYTES"; }
+  SUBCASE("Raw Binary data") { output1.datatype_ = "INT32"; }
+
+  outputs->insert(std::make_pair(output1.name_, output1));
+  cb::Error status = dataloader.ReadDataFromDir(inputs, outputs, dir);
+  CHECK(status.IsOk() == true);
+
+  const uint8_t* data_ptr{nullptr};
+  size_t batch1_size;
+
+  dataloader.GetOutputData("OUTPUT1", 0, 0, &data_ptr, &batch1_size);
+  CHECK(data_ptr == nullptr);
+  CHECK(batch1_size == 0);
+}
+
+TEST_CASE(
+    "dataloader: ReadDataFromDir: Mismatching Input Data" *
+    doctest::description("Successfully reading input files but having a "
+                         "mismatch will result in an error being thrown"))
+{
+  MockDataLoader dataloader;
+
+  std::string datatype;
+  std::string expected_error_message;
+
+  std::shared_ptr<ModelTensorMap> inputs = std::make_shared<ModelTensorMap>();
+  std::shared_ptr<ModelTensorMap> outputs = std::make_shared<ModelTensorMap>();
+
+  ModelTensor input1 = TestDataLoader::CreateTensor("INPUT1");
+  ModelTensor output1 = TestDataLoader::CreateTensor("OUTPUT1");
+
+  std::string dir{"mocked_out"};
+
+  std::vector<char> char_data{'0', '0', '0', '7', '5'};
+
+  std::vector<std::string> string_data{"InStr", "ExtraStr"};
+
+  SUBCASE("BYTES (string) data")
+  {
+    datatype = "BYTES";
+    EXPECT_CALL(dataloader, ReadTextFile(testing::_, testing::_))
+        .WillOnce(testing::DoAll(
+            testing::SetArgPointee<1>(string_data),
+            testing::Return(cb::Error::Success)));
+
+    SUBCASE("Dynamic shape")
+    {
+      input1.shape_ = {-1};
+      expected_error_message =
+          "input INPUT1 contains dynamic shape, provide shapes to send along "
+          "with the request";
+    }
+    SUBCASE("Supplied shape")
+    {
+      expected_error_message =
+          "provided data for input INPUT1 has 2 elements, expect 1";
+    }
+  }
+  SUBCASE("Raw Binary data")
+  {
+    datatype = "INT32";
+    EXPECT_CALL(dataloader, ReadFile(testing::_, testing::_))
+        .WillOnce(testing::DoAll(
+            testing::SetArgPointee<1>(char_data),
+            testing::Return(cb::Error::Success)));
+    SUBCASE("Dynamic shape")
+    {
+      input1.shape_ = {-1};
+      expected_error_message =
+          "input INPUT1 contains dynamic shape, provide shapes to send along "
+          "with the request";
+    }
+    SUBCASE("Supplied shape")
+    {
+      expected_error_message =
+          "provided data for input INPUT1 has byte size 5, expect 4";
```

Comment:
Where did `4` come from? Because `INT32` is 4 bytes? and the char array supplies 5?

Reply:
That is correct. Do you think I need to update the description or code in a way to make that more obvious?

Reply:
Sure. I reordered some things and added a comment for clarity.